src/or/onion.c

```diff
@@ -1339,3 +1342,466 @@ extended_cell_format(uint8_t *command_out, uint16_t *len_out,
   return 0;
 }
 
+/*****************************************************************************
+ *                          CREATE2V CELL HANDLING                           *
+ *****************************************************************************/
+
+/**
+ * Allocate memory and create a new <b>create2v_cell_t</b>.
+ *
+ * All variable length fields in the CREATE2V cell's body will be set
+ * to NULL, and all their corresponding lengths set to 0x00.
+ */
+create2v_cell_t*
+create2v_cell_new(void)
+{
+  create2v_cell_t *cell;
+
+  cell = tor_malloc(sizeof(create2v_cell_t));
+  cell->body = create2v_cell_body_new();
+
+  return cell;
+}
+
+/**
+ * Conduct various length checks on some handshake and padding data lengths for
+ * a given <b>handshake_type</b>.
+ *
+ * If <b>is_initiator</b> is true, then lengths are checked for the Alice side
+ * of the handshake, otherwise, if false, the checks are run for Bob's
+ * response.
+ */
+// XXXisis should this instead return bool and report errors rather than
+// asserting?
+static void
```

Comment:
Yes, that would be a good idea.

Reply:
And if this is going to be an assert at some point in the code, we should try to make it a nonfatal assertion if possible: those are less likely to turn into remote crash bugs.